### PR TITLE
change the way of killing cancelled tasks

### PR DESCRIPTION
**Pull request by tkumor3**
Closes #

## Introduced changes

-
-

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/forge/src/lib.rs [hunk 1]
```diff
@@ -552,7 +559,6 @@ fn run_with_fuzzing(
         let mut tasks = FuturesUnordered::new();
         // Pattern in order to waiting for things to finish shutting down
         // https://tokio.rs/tokio/topics/shutdown
```
Comment (cptartur) [inline crates/forge/src/lib.rs:561]: This comments should be moved/copied as well.

File: crates/forge/src/lib.rs [hunk 2]
```diff
@@ -441,6 +443,8 @@ async fn run_tests_from_crate(
         results.push(result);
     }
 
+    rec.close();
```
Comment (cptartur) [inline crates/forge/src/lib.rs:446]: Why do we close instead of dropping?

File: crates/forge/src/running.rs [hunk 3]
```diff
@@ -70,25 +71,17 @@ fn build_hints_dict<'b>(
...
     tokio::task::spawn_blocking(move || {
-        run_test_case(
-            args,
-            &case,
-            &runner,
-            &runner_config,
-            &runner_params,
-            &sender,
-        )
+        run_test_case(args, &case, &runner, &runner_config, &runner_params, &send)
```
Comment (cptartur) [inline crates/forge/src/running.rs:83]: Do we need to pass the sender down to the blocking task?

File: crates/forge/src/main.rs [hunk 4]
```diff
@@ -192,6 +193,7 @@ fn main_execution() -> Result<bool> {
             Ok(all_failed_tests)
         })
     })??;
+    rt.shutdown_background();
```
Comment (cptartur) [inline crates/forge/src/main.rs:196]: The docs say that this can cause resource leak because blocking tasks are not actually getting cancelled. https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.shutdown_background

**Review by cptartur — CHANGES_REQUESTED** [review]

File: crates/forge/src/lib.rs (hunk 2 above)
Comment (tkumor3) [reply crates/forge/src/lib.rs:446]: https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html - I want to close a channel. Not remove single sender

File: crates/forge/src/running.rs (hunk 3 above)
Comment (tkumor3) [reply crates/forge/src/running.rs:83]: I can refactor it a little bit

File: crates/forge/src/main.rs (hunk 4 above)
Comment (tkumor3) [reply crates/forge/src/main.rs:196]: I can remove it! It makes only the feeling that our tests finished a little bit faster! This task which is alive only checks if the chanel exists

File: crates/forge/src/lib.rs (hunk 1 above)
Comment (tkumor3) [reply crates/forge/src/lib.rs:561]: remove It use other pattern

File: crates/forge/src/running.rs [hunk 5]
```diff
@@ -70,25 +71,20 @@ fn build_hints_dict<'b>(
...
     tokio::task::spawn_blocking(move || {
-        run_test_case(
-            args,
-            &case,
-            &runner,
-            &runner_config,
-            &runner_params,
-            &sender,
-        )
+        if send.is_closed() {
```
Comment (Arcticae) [inline crates/forge/src/running.rs:86]: What if the channel is closed after we are past this section?

Comment (cptartur) [reply crates/forge/src/running.rs:86]: I think we'll get an error like previously. Perhaps this should be approached the other way round - "main" task should wait for all blocking tasks to terminate? Channel or some other messaging could be used for that (or maybe a simpler solution).

Comment (tkumor3) [reply crates/forge/src/running.rs:86]: in most cases, this test will be executed but user will not get the result/error. I think in extreme situations we will get >= x error like the previous, where x is the number of user cores.

Comment (tkumor3) [reply crates/forge/src/running.rs:87]: @Arcticae @cptartur  https://github.com/foundry-rs/starknet-foundry/pull/874/commits/792532dd89e4eaa10bc6ba5c4d67cea9a7bc7a2b

I fixed this problem

Comment (cptartur) [reply crates/forge/src/running.rs:92]: I think this should resolve the issue. We should add a comment somewhere in the code (probably in `run()`) explaining the logic of using two channels for this.

File: crates/forge/src/running.rs [hunk 6]
```diff
@@ -75,25 +76,31 @@ fn build_hints_dict<'b>(
...
-) -> Result<TestCaseSummary> {
+    send: Sender<()>,
+    send_shut_down: Sender<()>,
+) -> JoinHandle<Result<TestCaseSummary>> {
     tokio::task::spawn_blocking(move || {
+        // Due to the inability of spawn_blocking to be abruptly cancelled,
+        // a channel is used to receive information indicating
+        // that the execution of the task is no longer necessary.
+        if send.is_closed() {
+            return Err(anyhow::anyhow!("stop spawn_blocking"));
```
[the remark this answers is not in the document]
Comment (tkumor3) [reply crates/forge/src/running.rs:93]: No because this task has no parent

File: crates/forge/src/running.rs (hunk 5 above)
Comment (tkumor3) [reply crates/forge/src/running.rs:92]: 👍

**Review by cptartur — APPROVED** [review]

File: crates/forge/tests/integration/setup_fork.rs [hunk 7]
```diff
@@ -94,6 +94,7 @@ fn fork_aliased_decorator() {
     ).as_str());
 
     let rt = Runtime::new().expect("Could not instantiate Runtime");
+
```
Comment (MaksymilianDemitraszek) [inline crates/forge/tests/integration/setup_fork.rs:97]: Unrelated change

File: crates/forge/tests/integration/common/running_tests.rs [hunk 8]
```diff
@@ -12,6 +12,7 @@ use tokio::runtime::Runtime;
 
 pub fn run_test_case(test: &TestCase) -> Vec<TestCrateSummary> {
     let rt = Runtime::new().expect("Could not instantiate Runtime");
+
```
Comment (MaksymilianDemitraszek) [inline crates/forge/tests/integration/common/running_tests.rs:15]: Unrelated change

File: crates/forge/src/lib.rs [hunk 9]
```diff
@@ -364,7 +364,7 @@ pub async fn run(
         )
         .await?;
 
-        summaries.push(summary.clone());
+        summaries.push(summary);
```
Comment (MaksymilianDemitraszek) [inline crates/forge/src/lib.rs:367]: Unrelated change
